### PR TITLE
Provide a way to register / skip the shutdown hook in AutoConfiguredOpenTelemetrySdkBuilder

### LINKED ISSUE #4003: ShutdownHook in TracerProviderConfiguration
**Is your feature request related to a problem? Please describe.**
The `TracerProviderConfiguration` registers a ShutdownHook to close the `TracerProvider` (and shutdown the active span processor). In some environments, this may not be the most correct way to do it. Think for instance in a multi-app environment, each with its own `TracerProvider` and with different life cycles. The environments need to call `close` when the app terminates (or undeploys) and not on the JVM shutdown.

**Describe the solution you'd like**
Provide a way to skip the registration of the shutdown hook. This also means that we need to provide a wat to access the `SdkTracerProvider` to call `.close` (or the `Closeable` instance).  Right now this does not seem to be possible if you use auto-configuration.

### PR DESCRIPTION
**Pull request by radcortez**
- Fixes #4003

### CODE REVIEW

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java [hunk 1]
```diff
@@ -66,4 +66,18 @@ AutoConfigurationCustomizer addSpanExporterCustomizer(
...
+   * Control the registration of a shutdown hook to shut down the SDK when appropriate.
+   *
+   * <p>Skipping the registration of the shutdown hook may cause unexpected behavior. This
+   * configuration is for SDK consumers that require control over the SDK lifecycle. In this case,
+   * alternatives must be provided by the SDK consumer to shut down the SDK.
+   *
+   * @param registerShutdownHook a boolean <code>true</code> will register the hook, otherwise
+   *     <code>false</code> will skip registration.
+   */
+  default AutoConfigurationCustomizer registerShutdownHook(boolean registerShutdownHook) {
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:80]: Does this need to be defined in the interface? Is it insufficient as a method in `AutoConfiguredOpenTelemetrySdkBuilder` similar to `setResultAsGlobal(boolean setResultAsGlobal)`?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java [hunk 2]
```diff
@@ -197,6 +216,10 @@ public AutoConfiguredOpenTelemetrySdk build() {
             spanExporterCustomizer,
             samplerCustomizer);
 
+    if (registerShutdownHook) {
+      Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java:220]: `MeterProviderConfiguration` and `LogEmitterProviderConfiguration` also register shutdown hooks that should be conditional on `registerShutdownHook`.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java (hunk 1 above)
Comment (jack-berg) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:80]: Is there anyway to unit test this?

Comment (radcortez) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:80]: No, we can remove it from the interface. I've only added it here for consistency.

Comment (radcortez) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java:80]: There is no public API to query the registered shutdown hooks. One way would be to keep a reference to the `Thread` and then call `removeShutdownHook`.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java [hunk 3]
```diff
@@ -201,6 +219,14 @@ public AutoConfiguredOpenTelemetrySdk build() {
         LogEmitterProviderConfiguration.configureLogEmitterProvider(
             resource, config, meterProvider);
 
+    if (registerShutdownHook) {
+      Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));
```
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java:223]: Let's just use one thread, and call shutdown with `CompletableResultCode.ofAll().join` to reduce thread explosion on shutdown

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java [hunk 4]
```diff
@@ -201,6 +223,27 @@ public AutoConfiguredOpenTelemetrySdk build() {
...
+                    List<CompletableResultCode> shutdown = new ArrayList<>();
+                    shutdown.add(tracerProvider.shutdown());
+                    if (meterProvider instanceof SdkMeterProvider) {
+                      shutdown.add(((SdkMeterProvider) meterProvider).shutdown());
+                    }
+                    shutdown.add(logEmitterProvider.shutdown());
+                    CompletableResultCode.ofAll(shutdown).join(10, TimeUnit.SECONDS);
+                  }));
+
+      Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));
```
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java:240]: Looks like forgot to delete the old code with three threads

Comment (radcortez) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java]: Ops. Sorry. Fixed now.

**Review by anuraaga — APPROVED** [review]

**Review by jkwatson — APPROVED** [review]

**Review by jack-berg — APPROVED** [review]

**Pull request merged**